src/rpc/rpc.cc

```diff
@@ -813,6 +814,27 @@ namespace rpc {
       });
   }
 
+  void client::register_metrics() {
+      namespace sm = seastar::metrics;
+      metrics::metric_groups m;
+      auto domain_l = sm::label("domain")(_options.metrics_domain);
+      auto id_label = sm::label("id");
+      // This label is the aggregation one and is not exported, so it can be
+      // any unique identifier of a socket that doesn't have to make any sense
+      // to the metrics requestor. The object in-memory address fits both
+      auto id_l = id_label(reinterpret_cast<uintptr_t>(this));
+      m.add_group("rpc_client", {
+          sm::make_gauge("count", [] { return 1; }, sm::description("Total number of clients"), { domain_l, id_l }).aggregate({ id_label }),
+          sm::make_counter("sent_messages", _stats.sent_messages, sm::description("Total number of messages sent"), { domain_l, id_l }).aggregate({ id_label }),
+          sm::make_counter("replied", _stats.replied, sm::description("Total number of responses received"), { domain_l, id_l }).aggregate({ id_label }),
+          sm::make_counter("exception_received", _stats.exception_received, sm::description("Total number of exceptional responses received"), { domain_l, id_l }).aggregate({ id_label }),
+          sm::make_counter("timeout", _stats.timeout, sm::description("Total number of timeout responses"), { domain_l, id_l }).aggregate({ id_label }),
+          sm::make_gauge("pending", [this] { return outgoing_queue_length(); }, sm::description("Number of queued outbound messages"), { domain_l, id_l }).aggregate({ id_label }),
+          sm::make_gauge("wait_reply", [this] { return _outstanding.size(); }, sm::description("Number of replies waiting for"), { domain_l, id_l }).aggregate({ id_label }),
+      });
```

Comment:
Does this mean that if we have 1000 clients for the same domain, then each metrics scrape will add those 1000 metrics? And that we'll have 1000 group objects?

Reply:
Yes, that was sort of @amnonh  requirement to piggy-back aggregation. First version introduced the domain explicitly (but didn't name it domain) but had to shuffle metrics layer to refresh domains' stats per-scraping (https://github.com/xemul/seastar/commits/br-rpc-metrics-old)

Reply:
I was against the changes in the metrics layer that were related to scraping. You can still get the same results with a timer or update two counters, one that updates the per group and one that updates the aggregated one and register only the aggregated counter.